### ISSUE TITLE
Errno::EADDRINUSE: Address already in use - bind(2) for "" port 1900

### ISSUE THREAD

**Issue by bhardin** [denniskuczynski/simple_upnp#5]
I have no idea how no one has ever posted this issue.

I'm using a library which requires simple_upnp. When I attempt to use it, I receive the following error message:

```ruby
Errno::EADDRINUSE: Address already in use - bind(2) for "" port 1900
```

Hoping that I could look at this library to rebind the port somewhere else, but it looks like it is hard coded. I know very little on UPnP, so I don't know if the port is standard or not.

I attempted to use this library to do the `Discovery.search()` suggested in the README and got the same error.

```ruby
>> require 'simple_upnp'
=> true
>> include_location_details = true
=> true
>> devices = SimpleUpnp::Discovery.search()
Errno::EADDRINUSE: Address already in use - bind(2) for "" port 1900
	from /Users/bhardin/.rvm/gems/ruby-2.1.5/gems/simple_upnp-0.0.5/lib/simple_upnp/discovery.rb:43:in `bind'
	from /Users/bhardin/.rvm/gems/ruby-2.1.5/gems/simple_upnp-0.0.5/lib/simple_upnp/discovery.rb:43:in `open_socket'
	from /Users/bhardin/.rvm/gems/ruby-2.1.5/gems/simple_upnp-0.0.5/lib/simple_upnp/discovery.rb:15:in `search'
	from (irb):3
	from /Users/bhardin/.rvm/rubies/ruby-2.1.5/bin/irb:11:in `<main>'
```

Can anyone help?

**Comment by denniskuczynski** [comment]
Yes. 1900 is the standard port for uPnP UDP messages.
http://www.upnp-hacks.org/upnp.html

Do you have something else listening on that port?  What library are you using?
The UDP socket that listens during discovery should close after listening for the configured number of seconds.

Let me know if that helps.

**Comment by bhardin** [comment]
I'm using this library: https://github.com/ballantyne/wemo

But, even if I just use simple_upnp I have the same issue.

I've used `lsof -i | grep 1900` and the Network Utility for mac and nothing seems bound to port 1900. Let me know if you need me to dump any output.

**Comment by GGist** [comment]
I know this issue is fairly old but I stumbled upon it while looking for some information for my own UPnP library.

I just wanted to let you know that in my experience, it is not necessary to bind to port 1900 if all you are doing is sending an M-SEARCH request and receiving responses. The only time you need to bind to port 1900 is when you are listening for M-SEARCH requests or NOTIFY messages because, for example, when you bind to something like 0.0.0.0:1900 and subscribe to multicast 239.255.255.250, you will get all messages that were addressed to 239.255.255.250:1900. Essentially the router uses the local port as the port to route messages to when using multicast.

This is not necessary when making a search request because you are only looking for search responses and all devices are required to respond to M-SEARCH requests via unicast which means it responds to the sender's socket address meaning you can use any port. In my own library I just bind to port 0 which tells the OS to give me it's next available port.

As for the issue that @bhardin was having, making the library bind to port 0 should fix your problem, but if you ever wanted to listen for M-SEARCH requests or NOTIFY messages, I know at least on Windows there are socket options like [SO_EXCLUSIVEADDRUSE](https://msdn.microsoft.com/en-us/library/windows/desktop/cc150667(v=vs.85).aspx) which prevent binding to a certain port, even if SO_REUSEADDR is used which it looks like this library is using. In that case all you can do is track down the application that is doing that and terminating it.

**Comment by denniskuczynski** [comment]
Thanks for the info @GGist.

@bhardin I just pushed out 1.0.0 which should fix the issue.
